Unify behavior of the uploaders

bl-nero opened the pull request:
This change is required to enable pending summary state on the UI

- Changes the behavior of uploaders to allow uploading pending summaries. It also makes all uploaders consistently fail to overwrite existing recordings and final summaries.
- Upgrades the `gofakes3` library to support the If-None-Match headers.
- Copies the `MemBuffer` utility from the enterprise code.

Followed up by https://github.com/gravitational/teleport.e/pull/7461

Updates https://github.com/gravitational/teleport.e/issues/7373

bl-nero committed:
Unify behavior of the uploaders

- Changes the behavior of uploaders to allow uploading pending
  summaries. It also makes all uploaders consistently fail to overwrite
  existing recordings and final summaries.
- Upgrades the gofakes3 library to support the If-None-Match headers.
- Copies the MemBuffer utility from the enterprise code.

tigrato commented on lib/events/membuffer.go:
```diff
@@ -0,0 +1,37 @@
+package events
```
header file

bl-nero commented on lib/events/membuffer.go:
Yes, that's the plan (see https://github.com/gravitational/teleport.e/pull/7461)

bl-nero commented on lib/events/filesessions/fileuploader.go:
```diff
@@ -145,26 +149,45 @@ func downloadFile(path string, writer events.RandomAccessWriter) error {
 // UploadThumbnail writes a session thumbnail to a local directory.
 func (l *Handler) UploadThumbnail(ctx context.Context, sessionID session.ID, reader io.Reader) (string, error) {
-	return uploadFile(l.thumbnailPath(sessionID), reader)
+	return uploadFile(l.thumbnailPath(sessionID), reader, false /* overwrite */)
```
Sounds good

bl-nero commented on lib/events/s3sessions/s3handler.go:
```diff
@@ -322,36 +322,52 @@ func (h *Handler) Close() error {
 		Body:   reader,
 	}
+	if !overwrite {
+		uploadInput.IfNoneMatch = aws.String("*")
```
No, we didn't. (If we did, the fake S3 tests would likely fail, as before I upgraded the `gofakes3` library, it didn't support it.) It may ring a bell because it's been always used in the Azure uploader.

Up until now, we solely relied on S3 versioning to make sure that any attempt to override file contents would have no effect. I'm leaving the reading logic alone here, as we still need to properly support everything that has been uploaded until now, but the new S3 uploader will fail on an overwrite attempt, just like the Azure one does.

zmb3 commented on lib/events/s3sessions/s3handler.go:
```diff
@@ -372,31 +388,35 @@ func (h *Handler) uploadFile(ctx context.Context, path string, reader io.Reader)
-	return h.downloadFile(ctx, h.summaryPath(sessionID), writer)
+	err := h.downloadFile(ctx, h.summaryPath(sessionID), writer, nil /* versionID */)
+	if trace.IsNotFound(err) {
+		return trace.Wrap(h.downloadFile(ctx, h.pendingSummaryPath(sessionID), writer, nil /* versionID */))
```
Is the pending summary removed when a summary completes?

If so, do we have a potential race here where the first download fails because the summary hasn't finished, and then the summary finishes (and the pending summary is removed) before we attempt to download the pending summary?

bl-nero committed:
Fix a deadlock in gcsessions.Handler

The deadlock occurred when the upload passed a certain number of parts,
in which case the errors channel buffer got saturated and the object
deletion goroutines started piling up, ultimately leading to locking the
main loop that iterated through the object slice when the semaphore
channel got its buffer filled up too.

This deadlock prevented running the GCS uploader integration tests.

bl-nero committed:
Merge branch 'master' into bl-nero/pending-summaries

bl-nero committed:
review

tigrato reviewed (approved):
one nit that I think should be considered

bl-nero commented on lib/events/filesessions/fileuploader.go:
```diff
@@ -112,22 +118,26 @@ func (l *Handler) Close() error {
 func (l *Handler) DownloadSummary(ctx context.Context, sessionID session.ID, writer events.RandomAccessWriter) error {
-	return downloadFile(l.summaryPath(sessionID), writer)
+	err := downloadFile(l.summaryPath(sessionID), writer)
+	if trace.IsNotFound(err) {
```
Hah, that's actually a good idea. I'll split it into two functions then, `DonwloadSummary` and `DonwloadPendingSummary`.

bl-nero committed:
Another review round

- Move the pending session downloading to the caller
- Add a missing MemBuffer test
- Tweaks to the memory uploader (required by the Enterprise change)

bl-nero committed:
Build fix

zmb3 commented on lib/events/azsessions/azsessions.go:
```diff
@@ -269,6 +269,12 @@ func (h *Handler) sessionBlob(sessionID session.ID) *blockblob.Client {
 	return h.session.NewBlockBlobClient(sessionName(sessionID))
 }
 
+// summaryBlob returns a BlockBlobClient for the blob of the a pending session
```
```suggestion
// pendingSummaryBlob returns a BlockBlobClient for the blob of the a pending session
```

zmb3 reviewed (approved):
Approved withs some suggestions.

I don't love that the pending summaries never get cleaned up - I think people are likely to get confused when they see thousands of "pending" entries over long periods of time.

bl-nero commented:
> Approved withs some suggestions.
> 
> I don't love that the pending summaries never get cleaned up - I think people are likely to get confused when they see thousands of "pending" entries over long periods of time.

@zmb3 Do you think that calling them "drafts" instead would be better?

bl-nero committed:
review

bl-nero committed:
Merge branch 'master' into bl-nero/pending-summaries

bl-nero committed:
Merge branch 'master' into bl-nero/pending-summaries

bl-nero committed:
Fix lint and integration tests

bl-nero committed:
Delete the pending summaries when uploading final, use versioning on s3

bl-nero committed:
Merge branch 'master' into bl-nero/pending-summaries

bl-nero committed:
Some comments

bl-nero committed:
license

bl-nero committed:
Merge branch 'master' into bl-nero/pending-summaries
